### ISSUE TITLE
Reproducible TOC-crashes in crereader 

### ISSUE THREAD

**Issue by ghost** [koreader/kindlepdfviewer#311]
File 1: http://rghost.net/40557914
File 2: http://rghost.net/40558165
The above files are quite readable, but any call of expandable TOC crashes the reader. The old non-expandable TOC with the same books works properly.

**Comment by tigran123** [comment]
Yes, I have discovered this on Plutarch's "Lives" as well (from lib.rus.ec), a few days ago. This is because there are plenty of fb2 files with very broken structure, i.e. presenting their TOC as a tree does not work. I was going to either move the expandable TOC code into PDF and DjVu (i.e. return crereader TOC handling to the simple "linear" state that works) OR try to workaround within the TOC code for such broken entities.

**Comment by tigran123** [comment]
What is interesting is that the standalone CoolReader3 (under Linux) presents the TOC for Plutarch "Sravnitelnye zhizneopisanija" as a tree correctly, i.e. internally workarounds the broken structure. So, there is a good chance that we shall workaround this as well, in the existing expandable TOC code.

**Comment by tigran123** [comment]
Between the "Retrieving TOC..." and the final crash there are plenty of messages from crengine about "node w/o final parent":

```
# # InfoMessage  Retrieving TOC... 1
2012/09/24 20:53:59.2474 DEBUG node w/o final parent: 1794..86807
2012/09/24 20:53:59.2474 DEBUG node w/o final parent: 1794..86807
2012/09/24 20:53:59.2475 DEBUG node w/o final parent: 1923..37596
2012/09/24 20:53:59.2475 DEBUG node w/o final parent: 1923..37596
2012/09/24 20:53:59.2476 DEBUG node w/o final parent: 37596..86807
....
lua config error: ./unireader.lua:1652: bad argument #1 to 'insert' (table expected, got nil)
```

In the function UniReader:fillToc() at line 1652 it is impossible to get nil as long as the depths array is really a valid depths array, i.e. represents a tree. In this instance it is NOT, so we have to workaround it somehow...

**Comment by tigran123** [comment]
Ah, I see, there are nodes of depth>1 but with NO PARENT, that is what it is!

**Comment by tigran123** [comment]
@NuPogodi I expected the problem to be a lot more complex (as I have dealt with the broken n-ary trees before), but you forced me to attack the issue "on the head" and solve it. And I thank you for it.

**Comment by tigran123** [comment]
Fixed by the pull request #310, closing the issue.
